### PR TITLE
SQIL 

### PR DESCRIPTION
**Pull request by Will-Nie**
Add SQIL, an imitation learning algorithm. The algorithm supports lunarlander, cartpole, pong, spaceinvader, Qbert with the following commands:
ding -m serial_sqil -c lunarlander_sqil_config.py -s 0
ding -m serial_sqil -c cartpole_sqil_config.py -s 0
ding -m serial_sqil -c pong_sqil_config.py -s 0
ding -m serial_sqil -c spaceinvader_sqil_config.py -s 0
ding -m serial_sqil -c qbert_sqil_config.py -s 0

With respect to sql, it supports envs such as cartpole, lunarlander, pong, spaceinvader,qbert.

### CODE REVIEW

File: ding/model/wrapper/model_wrappers.py [hunk 1]
```diff
@@ -240,6 +240,49 @@ def forward(self, *args, **kwargs):
         output['action'] = action
         return output
 
+class EpsGreedySampleWrapper_sql(IModelWrapper):
```
Comment (PaParaZz1) [inline ding/model/wrapper/model_wrappers.py:243]: don't mix up  hump nomenclature and underscore nomenclature

Comment (PaParaZz1) [inline ding/model/wrapper/model_wrappers.py:243]: don't mix up  hump nomenclature and underscore nomenclature

File: ding/policy/sql.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+            # (bool) Whether to use multi gpu
+            multi_gpu=False,
+            # How many updates(iterations) to train after collector's one collection.
+            # Bigger "update_per_collect" means bigger off-policy.
+            # collect data -> update policy-> collect data -> ...
+            update_per_collect=3,  # after the batch data come into the learner, train with the data for 3 times
+            batch_size=64,
+            learning_rate=0.001,
+            # (float) L2 norm weight for network parameters.
+            weight_decay=0.0,
```
Comment (PaParaZz1) [inline ding/policy/sql.py:47]: remove `weight decay` field in default config like other policy

File: ding/policy/sql.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        data_n = q_nstep_td_data(
+            q_value, target_q_value, data['action'], target_q_action, data['reward'], data['done'], data['weight']
+        )
+        value_gamma = data.get('value_gamma')
+        loss, td_error_per_sample, record_target_v = q_nstep_sql_td_error(
+            data_n, self._gamma, self._cfg.learn.alpha, nstep=self._nstep, value_gamma=value_gamma
+        )
+        record_target_v = record_target_v.mean()
+        '''
+        print('This is loss')
```
Comment (PaParaZz1) [inline ding/policy/sql.py:140]: remove these debug codes which is commented now

File: ding/rl_utils/td.py [hunk 4]
```diff
@@ -469,6 +469,76 @@ def qrdqn_nstep_td_error(
...
+    assert len(action.shape) == 1, action.shape
+    if weight is None:
+        weight = torch.ones_like(action)
+
+    batch_range = torch.arange(action.shape[0])
+    q_s_a = q[batch_range, action]
+    #target_q_s_a = next_n_q[batch_range, next_n_action]
+    target_v = alpha * torch.logsumexp(next_n_q / alpha, 1)    #target_v = alpha * torch.log(torch.sum(torch.exp(next_n_q / alpha), 1)) 
+    #print(target_v) 
+    target_v[target_v == float("Inf")] = 20
```
Comment (PaParaZz1) [inline ding/rl_utils/td.py:516]: can we remove this hardcode now? or comment them and add corresponding statement

File: dizoo/atari/config/serial/pong/pong_sqil_config.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        nstep=3,
+        discount_factor=0.99,
+        learn=dict(
+            update_per_collect=10,
+            batch_size=32,
+            learning_rate=0.0001,
+            target_update_freq=500,
+            alpha = 0.12
+        ),
+        collect=dict(n_sample=96, demonstration_info_path = '/home/SENSETIME/User/nerveX/dizoo/atari/config/serial/pong/This_is_demo_data_pong/ckpt/ckpt_best.pth.tar'), #Users should add their own path here (path should lead to a well-trained model)
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/pong/pong_sqil_config.py:33]: Don't add any information about developers in public code and config

File: ding/rl_utils/td.py (hunk 4 above)
Comment (Will-Nie) [reply ding/rl_utils/td.py:516]: 目前感觉问题不大, 还是要跑一些实验看一下

Comment (Will-Nie) [reply ding/rl_utils/td.py:518]: For an appropriate hyper-parameter alpha, these hardcodes can be removed.  However, algorithms may face the danger of explosion for other alphas

File: ding/entry/serial_entry_sqil.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+def serial_pipeline_sqil(
+        input_cfg: Union[str, Tuple[dict, dict]],
+        seed: int = 0,
+        env_setting: Optional[List[Any]] = None,
+        model: Optional[torch.nn.Module] = None,
+        max_iterations: Optional[int] = int(1e10),
+) -> 'Policy':  # noqa
+    """
+    Overview:
+        Serial pipeline entry.
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_sqil.py:29]: update the related comments, such as what we do in this function to implement SQIL

File: ding/entry/serial_entry_sqil.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+from ding.worker import BaseLearner, SampleCollector, BaseSerialEvaluator, AdvancedReplayBuffer
+from ding.utils import set_pkg_seed
+
+
+def serial_pipeline_sqil(
+        input_cfg: Union[str, Tuple[dict, dict]],
+        seed: int = 0,
+        env_setting: Optional[List[Any]] = None,
+        model: Optional[torch.nn.Module] = None,
+        max_iterations: Optional[int] = int(1e10),
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_sqil.py:25]: add an input argument named `expert_model`, which is convenient for specializing user expert model, and when it is `None`, set it as DQN by default.

File: ding/entry/serial_entry_sqil.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        # Evaluate policy performance
+        if evaluator.should_eval(learner.train_iter):
+            stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+            if stop:
+                break
+        # Collect data by default config n_sample/n_episode
+        new_data = collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+        expert_data = expert_collector.collect(train_iter=learner.train_iter, policy_kwargs={'eps': -1})
+        for i in range(len(new_data)):
+            new_data[i]['reward'] = torch.tensor([0.])
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_sqil.py:120]: need to consider tensor device when use cuda

File: ding/entry/cli.py [hunk 9]
```diff
@@ -52,7 +52,7 @@ def print_registry(ctx: Context, param: Option, value: str):
 @click.option(
     '-m',
     '--mode',
-    type=click.Choice(['serial', 'parallel', 'dist', 'eval']),
+    type=click.Choice(['serial', 'serial_sqil', 'parallel', 'dist', 'eval']),
```
Comment (PaParaZz1) [inline ding/entry/cli.py:55]: This PR can be merged!

**Review by PaParaZz1 — APPROVED** [review]
